pkg/webhook/cert/provisioner.go

```diff
@@ -32,62 +33,92 @@ import (
+	// The ClientConfig will be used to inject into other objects,
+	// e.g. MutatingWebhookConfiguration and ValidatingWebhookConfiguration.
+	ClientConfig *admissionregistrationv1beta1.WebhookClientConfig
+
 	once sync.Once
 }
 
-// Sync takes a runtime.Object which is expected to be either a MutatingWebhookConfiguration or
-// a ValidatingWebhookConfiguration.
-// It provisions certificate for each webhook in the webhookConfiguration, ensures the cert and CA are valid,
-// and not expiring. It updates the CABundle in the webhook configuration if necessary.
-func (cp *Provisioner) Sync(webhookConfiguration runtime.Object) error {
-	var err error
+// Provision provisions certificates for for the WebhookClientConfig.
+// It ensures the cert and CA are valid and not expiring.
+// It updates the CABundle in the webhookClientConfig if necessary.
+func (cp *Provisioner) Provision(dryrun bool) (bool, error) {
 	// Do the initialization for CertInput only once.
 	cp.once.Do(func() {
 		if cp.CertGenerator == nil {
 			cp.CertGenerator = &generator.SelfSignedCertGenerator{}
 		}
-		if cp.Client == nil {
-			cp.Client, err = client.New(config.GetConfigOrDie(), client.Options{})
-			if err != nil {
-				return
-			}
-		}
-		if cp.CertWriter == nil {
-			cp.CertWriter, err = writer.NewCertWriter(
-				writer.Options{
-					Client:        cp.Client,
-					CertGenerator: cp.CertGenerator,
-				})
-			if err != nil {
-				return
-			}
-		}
 	})
+
+	if cp.CertWriter == nil {
+		return false, errors.New("CertWriter need to be set")
+	}
+
+	dnsName, err := dnsNameFromClientConfig(cp.ClientConfig)
 	if err != nil {
-		return fmt.Errorf("failed to default the CertProvision: %v", err)
+		return false, err
 	}
 
-	// Deepcopy the webhook configuration object before invoking EnsureCerts,
-	// since EnsureCerts will modify the provided object.
-	cloned := webhookConfiguration.DeepCopyObject()
-	err = cp.CertWriter.EnsureCerts(webhookConfiguration)
+	certs, changed, err := cp.CertWriter.EnsureCert(dnsName, dryrun)
 	if err != nil {
-		return err
+		return false, err
+	}
+
+	if changed {
+		// Ensure the CA bundle in the webhook configuration has the signing CA.
+		caBundle := cp.ClientConfig.CABundle
+		caCert := certs.CACert
+		if !bytes.Contains(caBundle, caCert) {
+			cp.ClientConfig.CABundle = append(caBundle, caCert...)
+		}
+	}
+	return changed, nil
+}
+
+// Inject the ClientConfig to the objects.
+// It supports MutatingWebhookConfiguration and ValidatingWebhookConfiguration.
+func (cp *Provisioner) Inject(objs ...runtime.Object) error {
+	for i := range objs {
+		switch typed := objs[i].(type) {
+		case *admissionregistrationv1beta1.MutatingWebhookConfiguration:
+			injectForEachWebhook(cp.ClientConfig, typed.Webhooks)
+		case *admissionregistrationv1beta1.ValidatingWebhookConfiguration:
+			injectForEachWebhook(cp.ClientConfig, typed.Webhooks)
+		}
+	}
+	return cp.CertWriter.Inject(objs...)
+}
```

Comment:
How do I ensure that caller have called `Provision` before `Inject` ? or is it okay ?